src/index/ivf/ivf.cc

```diff
@@ -219,77 +203,51 @@ IvfIndexNode<T>::Train(const DataSet& dataset, const Config& cfg) {
-        }
-        if constexpr (std::is_same<faiss::IndexIVFScalarQuantizer, T>::value) {
+            index_ = std::make_shared<faiss::IndexIVFPQ>(qzr, dim, nlist, ivf_pq_cfg.m, nbits, metric.value());
+            index_->own_fields = true;
+            index_->train(rows, (const float*)data);
+        } else if constexpr (std::is_same<faiss::IndexIVFScalarQuantizer, T>::value) {
             const IvfSqConfig& ivf_sq_cfg = static_cast<const IvfSqConfig&>(cfg);
+            auto nlist = MatchNlist(rows, ivf_sq_cfg.nlist);
             qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = new (std::nothrow) T(qzr, dim, MatchNlist(dataset.GetRows(), ivf_sq_cfg.nlist),
-                                         faiss::QuantizerType::QT_8bit, metric.value());
-        }
-        if constexpr (std::is_same<faiss::IndexBinaryIVF, T>::value) {
+            index_ = std::make_shared<faiss::IndexIVFScalarQuantizer>(qzr, dim, nlist, faiss::QuantizerType::QT_8bit,
+                                                                      metric.value());
+            index_->own_fields = true;
+            index_->train(rows, (const float*)data);
+        } else if constexpr (std::is_same<faiss::IndexBinaryIVF, T>::value) {
             const IvfBinConfig& ivf_bin_cfg = static_cast<const IvfBinConfig&>(cfg);
+            auto nlist = MatchNlist(rows, ivf_bin_cfg.nlist);
             qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = new (std::nothrow) T(qzr, dim, MatchNlist(dataset.GetRows(), ivf_bin_cfg.nlist), metric.value());
-        }
-        if (qzr == nullptr || index == nullptr) {
-            if (qzr) {
-                delete qzr;
-            }
-            if (index) {
-                delete index;
-            }
-            LOG_KNOWHERE_WARNING_ << "memory malloc error.";
-            return Status::malloc_error;
+            index_ = std::make_shared<faiss::IndexBinaryIVF>(qzr, dim, nlist, metric.value());
+            index_->own_fields = true;
+            index_->train(rows, (const uint8_t*)data);
         }
     } catch (std::exception& e) {
         if (qzr) {
             delete qzr;
         }
-        if (index) {
-            delete index;
-        }
         LOG_KNOWHERE_WARNING_ << "faiss inner error, " << e.what();
         return Status::faiss_inner_error;
     }
 
-    auto data = dataset.GetTensor();
-    auto rows = dataset.GetRows();
-    try {
-        if constexpr (std::is_same<faiss::IndexBinaryIVF, T>::value) {
-            index->train(rows, (const uint8_t*)data);
-        } else {
-            index->train(rows, (const float*)data);
-        }
-    } catch (std::exception& e) {
-        delete qzr;
-        delete index;
-        LOG_KNOWHERE_WARNING_ << "faiss inner error, " << e.what();
-        return Status::faiss_inner_error;
-    }
-    if (this->index_) {
```

Comment:
if train failed, break the old index, maybe not a good choice.

Reply:
if train fail, "qzr" is deleted manually, "index_" is deleted by system